docs/en/reference/query-builder.rst

```diff
@@ -368,6 +368,39 @@ or QueryBuilder instances to one of the following methods:
         ->orderBy('field', 'DESC')
         ->setMaxResults(100);
 
+Common Table Expressions
+~~~~~~~~~~~
+
+To define Common Table Expressions (CTEs) that can be used in select query.
+
+* ``with(string $name, string|QueryBuilder $queryBuilder, ?array $columns = null)``
+
+.. code-block:: php
+
+    <?php
+
+    $cteQueryBuilder1
+        ->select('id')
+        ->from('table_a')
+        ->where('id = :id')
+        ->setParameter('id', 1);
+
+    $cteQueryBuilder2
+        ->select('id')
+        ->from('table_b');
+
+    $queryBuilder
+        ->with('cte_a', $cteQueryBuilder1)
+        ->with('cte_b', $cteQueryBuilder2)
+        ->select('id')
+        ->from('cte_b', 'b')
+        ->join('b', 'cte_a', 'a', 'a.id = b.id');
+
+Multiple CTEs can be defined by calling the with method multiple times.
+
+Parameters used in a CTE can be defined directly in the QueryBuilder of the CTE.
```

Comment:
I'd replace "can" with "should" (we may even want to enforce this in the future).

Otherwise, it's not clear what will happen if a query uses multiple CTEs, of which some bind their own parameters but the others doesn't. Quite likely, there may be even misalignment of positional parameters.